File: dl3/android/DeepLabV3Demo/app/src/androidTest/java/org/pytorch/executorchexamples/dl3/UIWorkflowTest.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

package org.pytorch.executorchexamples.dl3;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.isEnabled;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static org.hamcrest.Matchers.not;
import static org.junit.Assert.assertTrue;

import android.graphics.drawable.BitmapDrawable;
import android.widget.ImageView;
import androidx.test.core.app.ActivityScenario;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.filters.LargeTest;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicReference;
import org.junit.Test;
import org.junit.runner.RunWith;

/**
 * UI workflow test that simulates the image segmentation workflow.
 *
 * Prerequisites:
 * - Push a .pte model file to app's private storage or ensure model can be downloaded
 *
 * This test validates:
 * - Model download/loading workflow
 * - Next sample image button functionality
 * - Run segmentation button functionality
 * - Reset image button functionality
 * - Inference time display after segmentation
 *
 * Note: This test assumes the model is already present or can be downloaded automatically.
 * For CI/CD, you may need to push the model file before running tests:
 * adb push dl3_xnnpack_fp32.pte /data/local/tmp/
 */
@RunWith(AndroidJUnit4.class)
@LargeTest
public class UIWorkflowTest {

    /**
     * Tests the basic UI elements are displayed on launch.
     * Verifies that:
     * 1. ImageView is displayed
     * 2. Download button is visible
     * 3. Control buttons visibility depends on model availability
     */
    @Test
    public void testInitialUIState() throws Exception {
        try (ActivityScenario<MainActivity> scenario = ActivityScenario.launch(MainActivity.class)) {
            // Wait for activity to fully load
            Thread.sleep(1000);

            // Verify ImageView is displayed
            onView(withId(R.id.imageView)).check(matches(isDisplayed()));

            // Verify download button exists
            onView(withId(R.id.downloadModelButton)).check(matches(isDisplayed()));

            // Progress bar should be invisible initially
            onView(withId(R.id.progressBar)).check(matches(not(isDisplayed())));
        }
    }

    /**
     * Tests the "Next" sample image button functionality.
     * Verifies that:
     * 1. Next button is clickable (when model is loaded)
     * 2. Clicking changes the displayed image
     * 3. Image changes multiple times when clicked repeatedly
     */
    @Test
    public void testNextSampleImage() throws Exception {
        try (ActivityScenario<MainActivity> scenario = ActivityScenario.launch(MainActivity.class)) {
            // Wait for activity to fully load
            Thread.sleep(1000);

            // Check if model is already loaded (Next button visible)
            AtomicBoolean modelLoaded = new AtomicBoolean(false);
            scenario.onActivity(activity -> {
                modelLoaded.set(activity.findViewById(R.id.nextButton).getVisibility() == android.view.View.VISIBLE);
            });

            if (!modelLoaded.get()) {
                // Model not loaded, try to load it first
                boolean downloadSuccessful = downloadOrLoadModel(scenario);
                if (!downloadSuccessful) {
                    // Skip test if model cannot be loaded
                    android.util.Log.i("UIWorkflowTest", "Skipping testNextSampleImage: model not available");
                    return;
                }
            }

            // Verify Next button is visible and enabled
            onView(withId(R.id.nextButton)).check(matches(isDisplayed()));
            onView(withId(R.id.nextButton)).check(matches(isEnabled()));

            // Get the initial image properties (width, height, and first pixel for simple comparison)
            AtomicReference<String> initialImageInfo = new AtomicReference<>();
            scenario.onActivity(activity -> {
                ImageView imageView = activity.findViewById(R.id.imageView);
                if (imageView.getDrawable() instanceof BitmapDrawable) {
                    android.graphics.Bitmap bitmap = ((BitmapDrawable) imageView.getDrawable()).getBitmap();
                    // Store basic image properties as a string
                    initialImageInfo.set(bitmap.getWidth() + "x" + bitmap.getHeight() + "@" + bitmap.getPixel(0, 0));
                }
            });

            // Click next button
            onView(withId(R.id.nextButton)).perform(click());
            Thread.sleep(500);

            // Verify image changed by comparing properties
            AtomicBoolean imageChanged = new AtomicBoolean(false);
            scenario.onActivity(activity -> {
                ImageView imageView = activity.findViewById(R.id.imageView);
                if (imageView.getDrawable() instanceof BitmapDrawable) {
                    android.graphics.Bitmap bitmap = ((BitmapDrawable) imageView.getDrawable()).getBitmap();
                    String currentImageInfo = bitmap.getWidth() + "x" + bitmap.getHeight() + "@" + bitmap.getPixel(0, 0);
                    // Check if image properties changed (different sample loaded)
                    imageChanged.set(!currentImageInfo.equals(initialImageInfo.get()));
                }
            });

            assertTrue("Image should change when Next button is clicked", imageChanged.get());
        }
    }

    /**
     * Tests the Run segmentation workflow.
     * Verifies that:
     * 1. Run button is enabled when model is loaded
     * 2. Clicking Run button starts segmentation
     * 3. Progress indicator appears during processing
     * 4. Inference time is displayed after completion
     * 5. Reset button becomes enabled after segmentation
     */
    @Test
    public void testRunSegmentation() throws Exception {
        try (ActivityScenario<MainActivity> scenario = ActivityScenario.launch(MainActivity.class)) {
            // Wait for activity to fully load
            Thread.sleep(1000);

            // Ensure model is loaded
            AtomicBoolean modelLoaded = new AtomicBoolean(false);
            scenario.onActivity(activity -> {
                modelLoaded.set(activity.findViewById(R.id.xnnpackButton).getVisibility() == android.view.View.VISIBLE);
            });

            if (!modelLoaded.get()) {
                // Try to load model
                boolean downloadSuccessful = downloadOrLoadModel(scenario);
                if (!downloadSuccessful) {
                    android.util.Log.i("UIWorkflowTest", "Skipping testRunSegmentation: model not available");
                    return;
                }
            }

            // Verify Run button is visible and enabled
            onView(withId(R.id.xnnpackButton)).check(matches(isDisplayed()));
            onView(withId(R.id.xnnpackButton)).check(matches(isEnabled()));

            // Verify reset button is initially disabled
            onView(withId(R.id.resetImage)).check(matches(not(isEnabled())));

            // Click Run button
            onView(withId(R.id.xnnpackButton)).perform(click());
            Thread.sleep(500);

            // Progress bar should be visible during inference
            // (might be too fast to catch, so we don't enforce this)

            // Wait for segmentation to complete (max 10 seconds)
            boolean completed = waitForSegmentationComplete(scenario, 10000);
            assertTrue("Segmentation should complete within 10 seconds", completed);

            // Verify inference time is displayed
            onView(withId(R.id.inferenceTimeText)).check(matches(isDisplayed()));

            // Verify reset button is now enabled
            onView(withId(R.id.resetImage)).check(matches(isEnabled()));

            // Verify Run button is enabled again (ready for next run)
            onView(withId(R.id.xnnpackButton)).check(matches(isEnabled()));
        }
    }

    /**
     * Tests the Reset image button functionality.
     * Verifies that:
     * 1. Reset button is disabled initially
     * 2. Reset button becomes enabled after running segmentation
     * 3. Clicking Reset restores the original sample image
     * 4. Reset button becomes disabled again after reset
     */
    @Test
    public void testResetImage() throws Exception {
        try (ActivityScenario<MainActivity> scenario = ActivityScenario.launch(MainActivity.class)) {
            // Wait for activity to fully load
            Thread.sleep(1000);

            // Ensure model is loaded
            AtomicBoolean modelLoaded = new AtomicBoolean(false);
            scenario.onActivity(activity -> {
                modelLoaded.set(activity.findViewById(R.id.xnnpackButton).getVisibility() == android.view.View.VISIBLE);
            });

            if (!modelLoaded.get()) {
                boolean downloadSuccessful = downloadOrLoadModel(scenario);
                if (!downloadSuccessful) {
                    android.util.Log.i("UIWorkflowTest", "Skipping testResetImage: model not available");
                    return;
                }
            }

            // Verify reset button is initially disabled
            onView(withId(R.id.resetImage)).check(matches(not(isEnabled())));

            // Run segmentation first
            onView(withId(R.id.xnnpackButton)).perform(click());
            boolean completed = waitForSegmentationComplete(scenario, 10000);
            assertTrue("Segmentation should complete", completed);

            // Reset button should now be enabled
            onView(withId(R.id.resetImage)).check(matches(isEnabled()));

            // Click reset button
            onView(withId(R.id.resetImage)).perform(click());
            Thread.sleep(500);

            // Verify reset button is disabled again
            onView(withId(R.id.resetImage)).check(matches(not(isEnabled())));

            // Verify inference time is hidden after reset
            onView(withId(R.id.inferenceTimeText)).check(matches(not(isDisplayed())));
        }
    }

    /**
     * Tests the complete workflow: Next -> Run -> Reset.
     * Verifies the full user workflow of:
     * 1. Loading a sample image
     * 2. Navigating to next sample
     * 3. Running segmentation
     * 4. Resetting the image
     */
    @Test
    public void testCompleteWorkflow() throws Exception {
        try (ActivityScenario<MainActivity> scenario = ActivityScenario.launch(MainActivity.class)) {
            // Wait for activity to fully load
            Thread.sleep(1000);

            // Ensure model is loaded
            AtomicBoolean modelLoaded = new AtomicBoolean(false);
            scenario.onActivity(activity -> {
                modelLoaded.set(activity.findViewById(R.id.xnnpackButton).getVisibility() == android.view.View.VISIBLE);
            });

            if (!modelLoaded.get()) {
                boolean downloadSuccessful = downloadOrLoadModel(scenario);
                if (!downloadSuccessful) {
                    android.util.Log.i("UIWorkflowTest", "Skipping testCompleteWorkflow: model not available");
                    return;
                }
            }

            // Step 1: Click Next to change sample image
            onView(withId(R.id.nextButton)).perform(click());
            Thread.sleep(500);

            // Step 2: Run segmentation
            onView(withId(R.id.xnnpackButton)).perform(click());
            boolean completed = waitForSegmentationComplete(scenario, 10000);
            assertTrue("Segmentation should complete", completed);

            // Verify results are shown
            onView(withId(R.id.inferenceTimeText)).check(matches(isDisplayed()));
            onView(withId(R.id.resetImage)).check(matches(isEnabled()));

            // Step 3: Reset image
            onView(withId(R.id.resetImage)).perform(click());
            Thread.sleep(500);

            // Verify reset worked
            onView(withId(R.id.resetImage)).check(matches(not(isEnabled())));
            onView(withId(R.id.inferenceTimeText)).check(matches(not(isDisplayed())));

            // Step 4: Can run segmentation again
            onView(withId(R.id.xnnpackButton)).check(matches(isEnabled()));
        }
    }

    /**
     * Tests the model download button state.
     * Verifies that:
     * 1. Download button shows appropriate state based on model availability
     * 2. Button text updates during download (if download is needed)
     */
    @Test
    public void testModelDownloadButton() throws Exception {
        try (ActivityScenario<MainActivity> scenario = ActivityScenario.launch(MainActivity.class)) {
            // Wait for activity to fully load
            Thread.sleep(1000);

            // Download button should be visible
            onView(withId(R.id.downloadModelButton)).check(matches(isDisplayed()));

            // Check the button state - it should either be:
            // - "Download Model" (model not present)
            // - "Model Ready" (model already loaded)
            // - "Downloading..." (download in progress)
            AtomicReference<String> buttonText = new AtomicReference<>();
            scenario.onActivity(activity -> {
                android.widget.Button downloadButton = activity.findViewById(R.id.downloadModelButton);
                buttonText.set(downloadButton.getText().toString());
            });

            android.util.Log.i("UIWorkflowTest", "Download button text: " + buttonText.get());
            
            // Verify the button text is one of the expected states
            assertTrue("Download button should show valid state",
                    buttonText.get().equals("Download Model") ||
                    buttonText.get().equals("Model Ready") ||
                    buttonText.get().contains("Downloading"));
        }
    }

    /**
     * Helper method to attempt to download or load the model.
     * Returns true if model is successfully loaded, false otherwise.
     */
    private boolean downloadOrLoadModel(ActivityScenario<MainActivity> scenario) throws InterruptedException {
        // Check if download button is enabled
        AtomicBoolean downloadEnabled = new AtomicBoolean(false);
        AtomicReference<String> buttonText = new AtomicReference<>();
        
        scenario.onActivity(activity -> {
            android.widget.Button downloadButton = activity.findViewById(R.id.downloadModelButton);
            downloadEnabled.set(downloadButton.isEnabled());
            buttonText.set(downloadButton.getText().toString());
        });

        // If model is already ready, return true
        if (buttonText.get().equals("Model Ready")) {
            return true;
        }

        // If download button is available and enabled, click it
        if (downloadEnabled.get() && buttonText.get().equals("Download Model")) {
            onView(withId(R.id.downloadModelButton)).perform(click());
            
            // Wait for download to complete (max 60 seconds)
            boolean downloadComplete = waitForModelReady(scenario, 60000);
            return downloadComplete;
        }

        return false;
    }

    /**
     * Waits for the model to be ready (download complete and loaded).
     * 
     * @param scenario the activity scenario
     * @param timeoutMs maximum time to wait in milliseconds
     * @return true if model is ready, false if timeout
     */
    private boolean waitForModelReady(ActivityScenario<MainActivity> scenario, long timeoutMs) throws InterruptedException {
        long startTime = System.currentTimeMillis();
        while (System.currentTimeMillis() - startTime < timeoutMs) {
            AtomicBoolean modelReady = new AtomicBoolean(false);
            scenario.onActivity(activity -> {
                // Model is ready when Run button is visible and enabled
                android.view.View runButton = activity.findViewById(R.id.xnnpackButton);
                modelReady.set(runButton.getVisibility() == android.view.View.VISIBLE && runButton.isEnabled());
            });
            
            if (modelReady.get()) {
                return true;
            }
            Thread.sleep(1000); // Poll every second
        }
        return false;
    }

    /**
     * Waits for segmentation to complete by monitoring UI state.
     * 
     * @param scenario the activity scenario
     * @param timeoutMs maximum time to wait in milliseconds
     * @return true if segmentation completed, false if timeout
     */
    private boolean waitForSegmentationComplete(ActivityScenario<MainActivity> scenario, long timeoutMs) throws InterruptedException {
        long startTime = System.currentTimeMillis();
        while (System.currentTimeMillis() - startTime < timeoutMs) {
            AtomicBoolean completed = new AtomicBoolean(false);
            scenario.onActivity(activity -> {
                // Segmentation is complete when:
                // 1. Progress bar is invisible
                // 2. Run button is enabled again
                // 3. Inference time is visible
                android.view.View progressBar = activity.findViewById(R.id.progressBar);
                android.view.View runButton = activity.findViewById(R.id.xnnpackButton);
                android.view.View inferenceTime = activity.findViewById(R.id.inferenceTimeText);
                
                completed.set(
                    progressBar.getVisibility() != android.view.View.VISIBLE &&
                    runButton.isEnabled() &&
                    inferenceTime.getVisibility() == android.view.View.VISIBLE
                );
            });
            
            if (completed.get()) {
                return true;
            }
            Thread.sleep(200); // Poll every 200ms
        }
        return false;
    }
}

File: dl3/android/DeepLabV3Demo/app/src/androidTest/java/org/pytorch/executorchexamples/dl3/UIWorkflowTest.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

package org.pytorch.executorchexamples.dl3

import android.content.Context
import android.util.Log
import androidx.compose.ui.test.*
import androidx.compose.ui.test.junit4.createAndroidComposeRule
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.filters.LargeTest
import org.junit.After
import org.junit.Before
import org.junit.Ignore
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import java.io.File

/**
 * UI workflow tests for the Compose-based DL3 demo.
 * 
 * Tests include:
 * - Download button functionality with and without model present
 * - Model run/segmentation testing
 * - UI state management (Next, Reset buttons)
 * - Inference time display
 */
@RunWith(AndroidJUnit4::class)
@LargeTest
class UIWorkflowTest {

    companion object {
        private const val MODEL_FILENAME = "dl3_xnnpack_fp32.pte"
        private const val MODEL_BACKUP_FILENAME = "dl3_xnnpack_fp32.pte.backup"
        private const val TAG = "UIWorkflowTest"
        private const val DOWNLOAD_TIMEOUT_MS = 120000L // 2 minutes
        private const val INFERENCE_TIMEOUT_MS = 30000L // 30 seconds
    }

    @get:Rule
    val composeTestRule = createAndroidComposeRule<MainActivity>()

    private lateinit var context: Context
    private lateinit var modelPath: String
    private lateinit var backupPath: String

    @Before
    fun setUp() {
        context = ApplicationProvider.getApplicationContext()
        modelPath = "${context.filesDir.absolutePath}/$MODEL_FILENAME"
        backupPath = "${context.filesDir.absolutePath}/$MODEL_BACKUP_FILENAME"
        
        // Ensure model is downloaded for most tests
        ensureModelAvailable()
    }

    @After
    fun tearDown() {
        // Clean up backup file if exists
        val backupFile = File(backupPath)
        if (backupFile.exists()) {
            backupFile.delete()
        }
    }

    /**
     * Ensures model is available by downloading if needed.
     */
    private fun ensureModelAvailable() {
        val modelFile = File(modelPath)
        if (!modelFile.exists()) {
            Log.i(TAG, "Model not found, downloading...")
            // Click download button and wait for completion
            composeTestRule.onNodeWithTag("downloadButton").assertExists()
            composeTestRule.onNodeWithTag("downloadButton").performClick()
            
            // Wait for download to complete (button should disappear)
            composeTestRule.waitUntil(timeoutMillis = DOWNLOAD_TIMEOUT_MS) {
                composeTestRule.onAllNodesWithTag("downloadButton")
                    .fetchSemanticsNodes().isEmpty()
            }
        } else {
            Log.i(TAG, "Model already available at $modelPath")
        }
    }

    /**
     * Tests that the initial UI is displayed correctly when model is ready.
     */
    @Test
    fun testInitialUIWithModel() {
        // Model buttons should be visible
        composeTestRule.onNodeWithTag("nextButton").assertIsDisplayed()
        composeTestRule.onNodeWithTag("pickButton").assertIsDisplayed()
        composeTestRule.onNodeWithTag("runButton").assertIsDisplayed()
        composeTestRule.onNodeWithTag("resetButton").assertIsDisplayed()
        
        // Image should be displayed
        composeTestRule.onNodeWithTag("segmentationImage").assertExists()
        
        // Download button should not be visible
        composeTestRule.onNodeWithTag("downloadButton").assertDoesNotExist()
        
        // Reset button should be disabled initially
        composeTestRule.onNodeWithTag("resetButton").assertIsNotEnabled()
    }

    /**
     * Tests the download button functionality when model is not present.
     * Uses rename-test-restore pattern to simulate missing model.
     */
    @Test
    fun testDownloadButtonWhenModelMissing() {
        val modelFile = File(modelPath)
        val backupFile = File(backupPath)
        
        // Step 1: Rename existing model to backup
        if (modelFile.exists()) {
            modelFile.renameTo(backupFile)
            Log.i(TAG, "Renamed model to backup")
        }
        
        try {
            // Step 2: Restart activity to show download button
            composeTestRule.activityRule.scenario.recreate()
            
            // Wait for UI to settle
            Thread.sleep(1000)
            
            // Download button should be visible
            composeTestRule.onNodeWithTag("downloadButton").assertIsDisplayed()
            composeTestRule.onNodeWithTag("downloadButton").assertIsEnabled()
            
            // Model buttons should not be visible
            composeTestRule.onNodeWithTag("nextButton").assertDoesNotExist()
            composeTestRule.onNodeWithTag("runButton").assertDoesNotExist()
            
            // Step 3: Click download button
            composeTestRule.onNodeWithTag("downloadButton").performClick()
            
            // Progress indicator should appear
            composeTestRule.onNodeWithTag("progressIndicator").assertExists()
            
            // Step 4: Wait for download to complete
            composeTestRule.waitUntil(timeoutMillis = DOWNLOAD_TIMEOUT_MS) {
                composeTestRule.onAllNodesWithTag("downloadButton")
                    .fetchSemanticsNodes().isEmpty()
            }
            
            // Model buttons should now be visible
            composeTestRule.onNodeWithTag("nextButton").assertIsDisplayed()
            composeTestRule.onNodeWithTag("runButton").assertIsDisplayed()
            
            // Verify new model file exists
            assert(modelFile.exists()) { "Model should be downloaded" }
            
        } finally {
            // Step 5: Restore backup if test downloaded new model and backup exists
            if (backupFile.exists()) {
                if (modelFile.exists()) {
                    modelFile.delete()
                }
                backupFile.renameTo(modelFile)
                Log.i(TAG, "Restored model from backup")
            }
        }
    }

    /**
     * Tests the "Next" button functionality to cycle through sample images.
     */
    @Test
    fun testNextButtonCyclesSamples() {
        // Click Next button
        composeTestRule.onNodeWithTag("nextButton").performClick()
        
        // Wait for image to change
        Thread.sleep(500)
        
        // Image should still be displayed
        composeTestRule.onNodeWithTag("segmentationImage").assertExists()
        
        // Can click Next again
        composeTestRule.onNodeWithTag("nextButton").performClick()
        Thread.sleep(500)
        composeTestRule.onNodeWithTag("segmentationImage").assertExists()
    }

    /**
     * Tests running segmentation and verifying inference time display.
     */
    @Test
    fun testRunSegmentation() {
        // Run button should be enabled
        composeTestRule.onNodeWithTag("runButton").assertIsEnabled()
        
        // Click Run button
        composeTestRule.onNodeWithTag("runButton").performClick()
        
        // Progress indicator should appear briefly
        // (might be too fast to catch consistently)
        
        // Wait for inference to complete
        composeTestRule.waitUntil(timeoutMillis = INFERENCE_TIMEOUT_MS) {
            composeTestRule.onAllNodesWithTag("inferenceTime")
                .fetchSemanticsNodes().isNotEmpty()
        }
        
        // Inference time should be displayed
        composeTestRule.onNodeWithTag("inferenceTime").assertIsDisplayed()
        
        // Reset button should now be enabled
        composeTestRule.onNodeWithTag("resetButton").assertIsEnabled()
        
        // Run button should still be enabled for next run
        composeTestRule.onNodeWithTag("runButton").assertIsEnabled()
    }

    /**
     * Tests the Reset button functionality.
     */
    @Test
    fun testResetButton() {
        // First run segmentation to enable reset
        composeTestRule.onNodeWithTag("runButton").performClick()
        
        composeTestRule.waitUntil(timeoutMillis = INFERENCE_TIMEOUT_MS) {
            composeTestRule.onAllNodesWithTag("inferenceTime")
                .fetchSemanticsNodes().isNotEmpty()
        }
        
        // Reset button should be enabled
        composeTestRule.onNodeWithTag("resetButton").assertIsEnabled()
        
        // Click Reset button
        composeTestRule.onNodeWithTag("resetButton").performClick()
        
        // Wait for reset to complete
        Thread.sleep(500)
        
        // Reset button should be disabled again
        composeTestRule.onNodeWithTag("resetButton").assertIsNotEnabled()
        
        // Inference time should disappear
        composeTestRule.onNodeWithTag("inferenceTime").assertDoesNotExist()
    }

    /**
     * Tests the complete workflow: Next -> Run -> Reset.
     */
    @Test
    fun testCompleteWorkflow() {
        // Step 1: Click Next to change sample
        composeTestRule.onNodeWithTag("nextButton").performClick()
        Thread.sleep(500)
        
        // Step 2: Run segmentation
        composeTestRule.onNodeWithTag("runButton").performClick()
        
        composeTestRule.waitUntil(timeoutMillis = INFERENCE_TIMEOUT_MS) {
            composeTestRule.onAllNodesWithTag("inferenceTime")
                .fetchSemanticsNodes().isNotEmpty()
        }
        
        // Verify results are shown
        composeTestRule.onNodeWithTag("inferenceTime").assertIsDisplayed()
        composeTestRule.onNodeWithTag("resetButton").assertIsEnabled()
        
        // Step 3: Reset image
        composeTestRule.onNodeWithTag("resetButton").performClick()
        Thread.sleep(500)
        
        // Verify reset worked
        composeTestRule.onNodeWithTag("resetButton").assertIsNotEnabled()
        composeTestRule.onNodeWithTag("inferenceTime").assertDoesNotExist()
        
        // Step 4: Can run segmentation again
        composeTestRule.onNodeWithTag("runButton").assertIsEnabled()
    }

    /**
     * Tests multiple consecutive runs to ensure model can be reused.
     * 
     * Note: Disabled due to known issue.
     */
    @Ignore("Known issue - test not working")
    @Test
    fun testMultipleConsecutiveRuns() {
        for (i in 1..3) {
            Log.i(TAG, "Running segmentation iteration $i")
            
            // Run segmentation
            composeTestRule.onNodeWithTag("runButton").performClick()
            
            composeTestRule.waitUntil(timeoutMillis = INFERENCE_TIMEOUT_MS) {
                composeTestRule.onAllNodesWithTag("inferenceTime")
                    .fetchSemanticsNodes().isNotEmpty()
            }
            
            // Verify inference time is displayed
            composeTestRule.onNodeWithTag("inferenceTime").assertIsDisplayed()
            
            // Reset for next iteration (except last)
            if (i < 3) {
                composeTestRule.onNodeWithTag("resetButton").performClick()
                Thread.sleep(500)
            }
        }
    }
}
